usl_pipeline/cloud_functions/main.py

```diff
@@ -287,6 +291,202 @@ def build_feature_matrix(cloud_event: functions_framework.CloudEvent) -> None:
+    """
+    # CityCAT output file paths are of the form:
+    # <study_area_name>/<config/path>/R11_C1_T<x>_y>min.rsl
+    study_area_name = blob_path.parts[0]
+    config_path = str(pathlib.PurePosixPath(*blob_path.parts[1:-1]))
+    logging.info(
+        "Processing file %s for study area %s simulation config %s",
+        blob,
+        study_area_name,
+        config_path,
+    )
+
+    # Enter the simulation in our metastore.
+    db = firestore.Client()
+    metastore.Simulation(
+        gcs_prefix_uri=f"gs://{blob.bucket.name}/{blob_path.parent}",
+        simulation_type=metastore.SimulationType.CITY_CAT,
+        study_area=metastore.StudyArea.get_ref(db, study_area_name),
+        configuration=metastore.FloodScenarioConfig.get_ref(db, config_path),
+    ).set(db)
+
+    # Retrieve geography information for the study area to rasterize CityCAT results.
+    header = metastore.StudyArea.get(db, study_area_name).as_header()
+
+    chunk_prefix = pathlib.PurePosixPath("chunks") / blob_path.parent
+    timestep = _timestep_number_from_rsl_path(blob_path.name)
+    logging.info(
+        "Writing chunks of %s for timestep %s to prefix %s",
+        blob,
+        timestep,
+        chunk_prefix,
+    )
+
+    with blob.open("rt") as citycat_result:
+        depths = simulation_readers.read_city_cat_result_as_raster(
+            citycat_result, header
+        )
+    logging.info("%s loaded, now being chunked.", blob)
+
+    for x_index, y_index, chunk in raster_chunkers.split_raster_into_chunks(
+        chunk_size, depths
+    ):
+        # Write chunks to paths like <x_index>_<y_index>/<timestep>.npy so that chunks
+        # at every timestep for the same location can be found under the same
+        # <x_index>_<y_index>/ prefix
+        chunk_name = chunk_prefix / f"{x_index}_{y_index}" / f"{timestep}.npy"
+        with blob.bucket.blob(str(chunk_name)).open("wb") as chunk_file:
+            numpy.save(chunk_file, chunk)
+
+    logging.info("Processed all chunks for file %s", blob)
+
+
+def _collapse_city_cat_output_chunks(
+    blob: storage.Blob, blob_path: pathlib.PurePosixPath, labels_bucket: storage.Bucket
+) -> None:
+    """Collects chunks for all timesteps at the same region and writes them as labels.
+
+    When a chunk file is uploaded, see if other chunks for the same region have been
+    uploaded for all timesteps. If so, collapse them into a single 3D tensor and write
+    them into the labels bucket.
+
+    See the documentation for process_citycat_outputs for more details.
+
+    Args:
+      blob: The blob containing the chunk .npy file.
+      blob_path: The path to the blob within its bucket.
+      labels_bucket: The bucket in which to write the label tensor.
+    """
+    # Chunks are of the form
+    # chunks/<study_area_name>/<config/path>/<x_index>_<y_index>/<timestep>.npy
+    study_area_name = blob_path.parts[1]
+    config_path = str(pathlib.PurePosixPath(*blob_path.parts[2:-2]))
```

Comment:
Can you please comment here that config_path may consist of several folders?

Reply:
done